// @what3words/api/dist/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.what3words = void 0;
var client_1 = require("./client");
function what3words(apiKey, config, opts) {
    var transport = (opts === null || opts === void 0 ? void 0 : opts.transport) || require('./lib').fetchTransport();
    var autosuggestClient = new client_1.AutosuggestClient(apiKey, config, transport);
    var availableLanguagesClient = new client_1.AvailableLanguagesClient(apiKey, config, transport);
    var convertTo3waClient = new client_1.ConvertTo3waClient(apiKey, config, transport);
    var convertToCoordinatesClient = new client_1.ConvertToCoordinatesClient(apiKey, config, transport);
    var gridSectionClient = new client_1.GridSectionClient(apiKey, config, transport);
    var service = {
        clients: {
            autosuggest: autosuggestClient,
            availableLanguages: availableLanguagesClient,
            convertTo3wa: convertTo3waClient,
            convertToCoordinates: convertToCoordinatesClient,
            gridSection: gridSectionClient,
        },
        setApiKey: function (apiKey) {
            autosuggestClient.apiKey(apiKey);
            availableLanguagesClient.apiKey(apiKey);
            convertTo3waClient.apiKey(apiKey);
            convertToCoordinatesClient.apiKey(apiKey);
            gridSectionClient.apiKey(apiKey);
        },
        setConfig: function (config) {
            autosuggestClient.config(config);
            availableLanguagesClient.config(config);
            convertTo3waClient.config(config);
            convertToCoordinatesClient.config(config);
            gridSectionClient.config(config);
        },
        autosuggest: autosuggestClient.run.bind(autosuggestClient),
        autosuggestSelection: autosuggestClient.onSelected.bind(autosuggestClient),
        availableLanguages: availableLanguagesClient.run.bind(availableLanguagesClient),
        convertTo3wa: convertTo3waClient.run.bind(convertTo3waClient),
        convertToCoordinates: convertToCoordinatesClient.run.bind(convertToCoordinatesClient),
        gridSection: gridSectionClient.run.bind(gridSectionClient),
    };
    return service;
}
exports.what3words = what3words;
